pkg/controller/endpoint/endpoints_controller_cache.go

```diff
@@ -0,0 +1,232 @@
+
+	// Mutex guarding this util.
+	mutex sync.Mutex
+}
+
+// Creates new instance of the endpointsControllerCache.
+func newEndpointsControllerCache() *endpointsControllerCache {
+	return &endpointsControllerCache{
+		pods:     make(map[ /* namespace */ string]map[ /* name */ string]*v1.Pod),
+		services: make(map[ /* namespace */ string]map[ /* name */ string]*v1.Service),
+	}
+}
+
+// Initialize initializes the cache with pods and services obtained from the provided listers.
+// Pods and services are listed behind mutex, which temporarily disables event listener callbacks
+// (e.g. AddPod, UpdateService, etc.). It guarantees that no information will be lost.
+// This method should be called exactly once on the startup, before the first call of the
+// GetServiceAndPods method, after the lister have synced. Please note that this initialization
+// scheme relies on the assumption that, when a handler is registered in the Informer, the current
+// state will be fully delivered to the event handler via the OnAdd callbacks.
+// TODO(mm4tt): Support initialization natively in the Informer framework and get rid of this
+// 							method. This will require a new signal (and callback in the event handler) informing
+//              that the state has been synced and fully delivered to the event handlers via the
+//              OnAdd callbacks.
+func (e *endpointsControllerCache) Initialize(serviceLister corelisters.ServiceLister, podLister corelisters.PodLister) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	services, err := serviceLister.List(labels.Everything())
+	if err != nil {
+		// Error shouldn't happened as we're getting objects from a local cache, so fail if it does.
+		klog.Fatalf("unable to list services due to %v", err)
+	}
+	// Please note that we're listing pods at a different time than services, so it's possible that we
+	// will initialize the cache with a pod state newer than the service state. This inaccuracy
+	// doesn't matter in the long run though. If the service state is older and doesn't reflect what
+	// is stored in etcd when the pods are listed, then we're guaranteed to/ get service state updates
+	// via the (Add/Update/Delete)Service methods which will eventually make the state consistent.
+	pods, err := podLister.List(labels.Everything())
+	if err != nil {
+		// Error shouldn't happened as we're getting objects from a local cache, so fail if it does.
+		klog.Fatalf("unable to list pods due to %v", err)
+	}
+
+	for _, service := range services {
+		e.getServices(service.Namespace)[service.Name] = service
+	}
+	for _, pod := range pods {
+		e.getPods(pod.Namespace)[pod.Name] = pod
+	}
+}
+
+// AddPod adds new pod to the cache and returns a list of namespaced names of service / endpoints
+// objects that should be synced after this operation has been completed.
+func (e *endpointsControllerCache) AddPod(pod *v1.Pod) []string {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	e.getPods(pod.Namespace)[pod.Name] = pod
+
```

Comment:
nit: remove this empty line maybe (same below)?
Having 2 empty lines out of 6 lines looks a bit strange to me :-)